### PR TITLE
Swa 96 use experiment as main interface for user to interact with

### PR DESCRIPTION
**Pull request by layflo**
- create experiment using config file
- add setup_model_trainer function with checkpoint and tensorboard callbacks as default
- add start_training function 

Still left: 
- Serialization of model trainer(s) after training
- Pause training
- Resume training
- evaluate

### CODE REVIEW

File: sonosco/training/experiment.py [hunk 1]
```diff
@@ -27,13 +38,15 @@ class Experiment:
...
                  experiment_name,
+                 seed:int = None,
                  experiments_path=None,
                  sub_directories=("plots", "logs", "code", "checkpoints"),
                  exclude_dirs=('__pycache__', '.git', 'experiments'),
                  exclude_files=('.pyc',)):
 
         self.experiments_path = self._set_experiments_dir(experiments_path)
         self.name = self._set_experiment_name(experiment_name)
+        if seed is not None: self._set_seed(seed)
```
Comment (yuriyarabskyy) [inline sonosco/training/experiment.py]: should be in separate lines

File: sonosco/training/experiment.py [hunk 2]
```diff
@@ -99,12 +119,78 @@ def add_directory(self, dir_name):
...
+        trainer = ModelTrainer(model=model, loss=loss, epochs=epochs,
+                               train_data_loader=train_data_loader,
+                               val_data_loader=val_data_loader, decoder=decoder,
+                               optimizer=optimizer, lr=lr, custom_model_eval=custom_model_eval,
+                               gpu=gpu, clip_grads=clip_grads,
+                               metrics=metrics, callbacks=callbacks)
+        if model_checkpoints:
+            date_time = datetime.datetime.fromtimestamp(time()).strftime('%Y-%m-%d_%H:%M:%S')
+            model_name = name + '_' + date_time + '_checkpoint.pt'
+            trainer.add_callback(ModelCheckpoint(output_path=self.logs, model_name=model_name))
```
Comment (yuriyarabskyy) [inline sonosco/training/experiment.py]: I created a different directory for checkpoints, self.checkpoints

File: sonosco/training/experiment.py [hunk 3]
```diff
@@ -99,12 +119,78 @@ def add_directory(self, dir_name):
         # create directory
         path_utils.try_create_directory(dir_path)
 
+    def setup_model_trainer(self,
```
Comment (yuriyarabskyy) [inline sonosco/training/experiment.py]: to not repeat all the arguments twice I would rather have here `*args, **kwargs` and pass those downwards to ModelTrainer

File: sonosco/training/experiment.py [hunk 4]
```diff
@@ -99,12 +119,78 @@ def add_directory(self, dir_name):
...
+                               optimizer=optimizer, lr=lr, custom_model_eval=custom_model_eval,
+                               gpu=gpu, clip_grads=clip_grads,
+                               metrics=metrics, callbacks=callbacks)
+        if model_checkpoints:
+            date_time = datetime.datetime.fromtimestamp(time()).strftime('%Y-%m-%d_%H:%M:%S')
+            model_name = name + '_' + date_time + '_checkpoint.pt'
+            trainer.add_callback(ModelCheckpoint(output_path=self.logs, model_name=model_name))
+
+        if tensorboard:
+            trainer.add_callback(TensorBoardCallback(log_dir=self.logs))
```
Comment (yuriyarabskyy) [inline sonosco/training/experiment.py]: there is a plots directory in the experiment class

File: sonosco/training/experiment.py [hunk 5]
```diff
@@ -99,12 +119,78 @@ def add_directory(self, dir_name):
...
+
+        self.model_trainers = {name: trainer}
```
Comment (yuriyarabskyy) [inline sonosco/training/experiment.py]: why is it a dictionary?

File: sonosco/training/experiment.py [hunk 6]
```diff
@@ -99,12 +119,78 @@ def add_directory(self, dir_name):
...
+
+        trainer = ModelTrainer(model=model, loss=loss, epochs=epochs,
+                               train_data_loader=train_data_loader,
+                               val_data_loader=val_data_loader, decoder=decoder,
+                               optimizer=optimizer, lr=lr, custom_model_eval=custom_model_eval,
+                               gpu=gpu, clip_grads=clip_grads,
+                               metrics=metrics, callbacks=callbacks)
+        if model_checkpoints:
+            date_time = datetime.datetime.fromtimestamp(time()).strftime('%Y-%m-%d_%H:%M:%S')
+            model_name = name + '_' + date_time + '_checkpoint.pt'
```
Comment (yuriyarabskyy) [inline sonosco/training/experiment.py]: this should be a default name inside of the checkpoint callback I think

File: sonosco/training/experiment.py [hunk 7]
```diff
@@ -99,12 +119,78 @@ def add_directory(self, dir_name):
...
+            trainer.add_callback(ModelCheckpoint(output_path=self.logs, model_name=model_name))
+
+        if tensorboard:
+            trainer.add_callback(TensorBoardCallback(log_dir=self.logs))
+
+        self.model_trainers = {name: trainer}
+
+    def start_training(self, trainer_name: str = None):
```
Comment (yuriyarabskyy) [inline sonosco/training/experiment.py]: Not sure if we need a start training here or maybe we should rather call experiment.trainer.start_training(). Or is there a reason why an experiment could have multiple trainers?

File: sonosco/training/trainer.py [hunk 8]
```diff
@@ -37,7 +37,7 @@ def __init__(self,
                  train_data_loader: DataLoader,
                  val_data_loader: DataLoader = None,
                  decoder = None,
-                 optimizer=torch.optim.Adam,
+                 optimizer = torch.optim.Adam,
```
Comment (yuriyarabskyy) [inline sonosco/training/trainer.py]: correct style is without spaces in the arguments before and after `=`

Comment (yuriyarabskyy) [reply sonosco/training/trainer.py]: unless you specify a type then you add spaces

**Comment by yuriyarabskyy** [comment]
also conflicts with develop should be resolved

File: sonosco/training/experiment.py (hunk 3 above)
Comment (layflo) [reply sonosco/training/experiment.py]: I thought about that aswell, but in my opinion, one should use this function for generally using the model trainer, and when calling this function, can one check what entries you can put there, when it is not specified in this function but just in the model trainer?

Also thought about deleting then the default values in the model trainer init.

File: sonosco/training/experiment.py (hunk 4 above)
Comment (layflo) [reply sonosco/training/experiment.py]: but this is just about the tensorboard log file, or? so there are not really plots saved, or?

File: sonosco/training/experiment.py (hunk 5 above)
Comment (layflo) [reply sonosco/training/experiment.py]: the idea was (discussed with wiktor)
that you maybe have multiple model trainers per experiment. For example for different models. So when you want to start training you can either choose a specific model trainer to start training (by its name) or all

File: sonosco/training/experiment.py (hunk 7 above)
Comment (layflo) [reply sonosco/training/experiment.py]: see above.

but of course we can discuss if one would even need multiple model trainers per experiment

File: sonosco/training/experiment.py (hunk 3 above)
Comment (layflo) [reply sonosco/training/experiment.py]: plus I am not sure how this **kwargs works :D

If I have named arguments in the function, and I pass **kwargs to the model trainer, are then the once that are just for the experiment function left out? or does this lead to an error when there are named arguments, that are not in the model trainer?

Comment (yuriyarabskyy) [reply sonosco/training/experiment.py]: My biggest argument for why repeating all the arguments is bad is that if you want to change something you always have to remember to change it at three places and this is definitely bad coding style then. The **kwargs to the model trainer would throw an error if you give some arguments which shouldn't be there unless ModelTrainer also has **kwargs in the end which picks up all the unmatched args.

File: sonosco/training/experiment.py (hunk 4 above)
Comment (yuriyarabskyy) [reply sonosco/training/experiment.py]: Yes, we have no plots but I would rather have the logs directory for the actual log files of the program. And for these tensorboard logs I would rather have a separate dir to not mix them up.

File: sonosco/training/experiment.py (hunk 5 above)
Comment (yuriyarabskyy) [reply sonosco/training/experiment.py]: I would say one experiment is for just one model training -> one model trainer. But we can discuss this later, np.

**Pull request merged**